Point cloud identify tool 2D

NEDJIMAbelgacem opened the pull request:
## Description

This PR makes 2D identification tool works with point clouds.

NEDJIMAbelgacem committed:
Initial implementation of selection tool for point cloud

NEDJIMAbelgacem committed:
Refactoe QgsPointCloudLayerRenderer::traverseTree

NEDJIMAbelgacem committed:
Add classified renderer filtering behaviour

NEDJIMAbelgacem committed:
add fix for point selection

NEDJIMAbelgacem committed:
change to variant and remove QgsPointCloudIdentifyResults

NEDJIMAbelgacem committed:
Add identify when rendering as circles

nyalldawson commented on src/core/pointcloud/qgspointcloudclassifiedrenderer.cpp:
```diff
@@ -129,6 +129,21 @@ void QgsPointCloudClassifiedRenderer::renderBlock( const QgsPointCloudBlock *blo
+  if ( !pointAttributes.contains( mAttribute ) )
+    return false;
+  bool parsedCorrectly;
+  int attributeInt = pointAttributes[ mAttribute ].toReal( &parsedCorrectly );
```
```suggestion
  int attributeInt = pointAttributes[ mAttribute ].toInt( &parsedCorrectly );
```

nyalldawson commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -172,13 +174,127 @@ bool QgsPointCloudLayerRenderer::forceRasterRender() const
 }
 
-QList<IndexedPointCloudNode> QgsPointCloudLayerRenderer::traverseTree( const QgsPointCloudIndex *pc,
+QVector<QMap<QString, QVariant>> QgsPointCloudLayerRenderer::identify( const QgsGeometry &geometry, const QgsIdentifyContext &identifyContext )
```
I would suggest splitting this function up into two parts:

1. A method in QgsPointCloudDataProvider which accepts a QgsBox3d extent (which must be in the layer's crs), a maximum screen error, and a maximum number of point to identify (or -1 for "every point"). This method would then contain all the index specific logic, and NOT handle any renderer related concepts.
2. A renderedPointsInBounds() method in QgsPointCloudRenderer (not QgsPointCloudLayerRenderer), which calls the method from (1) to get ALL the points from the data provider, and then does the point size and shape handling you've got here. This method would need arguments for the QgsBox3d, a render context, and the identify context.

This would be cleaner, and the main benefit is that we would then have a nice method for plugins to use to also perform a "give me the points at this position" search on a point cloud layer.

NEDJIMAbelgacem replied:
My reason behind implementing this in the layer renderer is because with have access to rendering informations like the screen errors and the classification stuff.
How can we deal with the level of detail and classification if we move it to the data provider ?

NEDJIMAbelgacem committed:
Merge branch 'master' into point-cloud-identify-tool

NEDJIMAbelgacem committed:
Address Nyall suggestions

NEDJIMAbelgacem committed:
fix spelling

NEDJIMAbelgacem committed:
add docs

NEDJIMAbelgacem committed:
fix unused variable

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -176,3 +181,94 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
+  const IndexedPointCloudNode root = index->root();
+
+
+  QgsPointCloudLayerElevationProperties *properties = dynamic_cast<QgsPointCloudLayerElevationProperties *>( layer->elevationProperties() );
```
I think we should remove the dependance on QgsPointCloudLayer here, and require that callers manually handle the elevation properties related logic themselves if desired.

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -176,3 +181,94 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
+
+QVector<QMap<QString, QVariant>> QgsPointCloudDataProvider::identify(
+                                QgsPointCloudLayer *layer,
+                                float maximumError,
```
A cleaner API would be to remove rootErrorPixels, and require that maximumError is specified in the data provider's native units. (I.e. require that the caller does the work to convert their desired maximum error into the correct units.) That would allow removal of rootErrorPixels, because it can be calculated in this method directly from the index root node.

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -176,3 +181,94 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
+  QgsPointCloudRequest request;
+  request.setAttributes( attributeCollection );
+
+  QgsPointCloudRenderContext context( renderContext, index->scale(), index->offset(), properties->zScale(), properties->zOffset() );
```
Can you remove this? It's only used during the calculation of the point x/y, and that can be directly calculated using the index's offset/scaling. QgsPointCloudRenderContext is a renderer specific class, and we shouldn't be using it outside of rendering.

nyalldawson commented on src/core/pointcloud/qgspointcloudrenderer.cpp:
```diff
@@ -190,4 +194,49 @@ void QgsPointCloudRenderer::setPointSymbol( PointSymbol symbol )
+
+  QVector<QMap<QString, QVariant>> points = layer->dataProvider()->identify( layer, maximumError, rootErrorPixels, geometry, QgsDoubleRange() );
+
+  for ( QMap<QString, QVariant> point : points )
+  {
+    if ( this->willRenderPoint( point ) )
+      selectedPoints.push_back( point );
+  }
```
A more modern c++ approach:

```suggestion
  points.erase( std::remove_if( points.begin(), points.end(), [this]( const QMap<QString, QVariant> &point )
  {
    return !willRenderPoint( point );
  } ), points.end() );
```

NEDJIMAbelgacem committed:
make single click identify work

NEDJIMAbelgacem committed:
remove layer and elevation properties

NEDJIMAbelgacem committed:
remove root error from identify in the provider

NEDJIMAbelgacem commented:
Hi @nyalldawson , I addressed your reviews but there is still one problem left: How do I convert the maximum error factor to the provider's native units?

NEDJIMAbelgacem commented:
Maybe @PeterPetrik can help me figure this out

NEDJIMAbelgacem committed:
Merge branch 'master' into point-cloud-identify-tool

NEDJIMAbelgacem committed:
Parralelize fetching points from IndexedPointCloudNode

NEDJIMAbelgacem committed:
fix banned keyword check

NEDJIMAbelgacem committed:
Add returned point clouds limit

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -176,3 +183,145 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
+  QgsPointCloudRequest request;
+  request.setAttributes( attributeCollection );
+
+  QgsPointCloudRenderContext context( renderContext, index->scale(), index->offset(), 1.0, 0.0 );
```
We still need to remove this from the data provider class -- it's a render specific class which should not be used here

nyalldawson commented on src/core/pointcloud/qgspointcloudlayerrenderer.h:
```diff
@@ -59,8 +64,8 @@ class CORE_EXPORT QgsPointCloudLayerRenderer: public QgsMapLayerRenderer
 
-    //! Traverses tree and returns all nodes in specified depth
-    QList<IndexedPointCloudNode> traverseTree( const QgsPointCloudIndex *pc, const QgsRenderContext &context, IndexedPointCloudNode n, float maxErrorPixels, float nodeErrorPixels );
+    //! Traverses tree and returns all nodes that intersects with a \a geometry in specified depth (if the geometry is null all nodes are considered)
```
As far as i can see, the new geometry option isn't actually used anywhere. Is this a remnant from an earlier version?

NEDJIMAbelgacem replied:
It is used here https://github.com/qgis/QGIS/blob/19afdbe2359159da871f7d22b252c766095a87d9/src/core/pointcloud/qgspointclouddataprovider.cpp#L236

nyalldawson commented on src/core/pointcloud/qgspointcloudrenderer.h:
```diff
@@ -189,6 +190,63 @@ class CORE_EXPORT QgsPointCloudRenderContext
     }
+
+    //! Returns a map containing the attributes of a certain point from the point cloud
+    QMap<QString, QVariant> attributeMap( const char *data, std::size_t recordOffset, const QgsPointCloudAttributeCollection &attributeCollection )
```
Same here -- this doesn't seem to be used anywhere anymore. Can it be removed now?

NEDJIMAbelgacem replied:
It is used here https://github.com/qgis/QGIS/blob/19afdbe2359159da871f7d22b252c766095a87d9/src/core/pointcloud/qgspointclouddataprovider.cpp#L236

NEDJIMAbelgacem replied:
Nevermind, I moved it to the data provider class

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -176,3 +183,145 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
+    QgsPointCloudAttributeCollection blockAttributes = block->attributes();
+    const std::size_t recordSize = blockAttributes.pointRecordSize();
+    mContext.setAttributes( block->attributes() );
+    for ( int i = 0; i < block->pointCount(); ++i )
```
```suggestion
    std::unique_ptr< QgsGeometryEngine > extentEngine( mExtentGeometry.constGet() );
    extentEngine->prepareGeometry();
    for ( int i = 0; i < block->pointCount(); ++i )    
```

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -176,3 +183,145 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
+      double x, y, z;
+      _pointXY( mContext, ptr, i, x, y );
+      z = _pointZ( mContext, ptr, i );
+      QgsPointXY pointXY( x, y );
+
+      if ( pointsCount < mPointsLimit && mExtentGeometry.contains( &pointXY ) && mZRange.contains( z ) )
```
```suggestion
      QgsPoint pointXY( x, y );

      if ( pointsCount < mPointsLimit && extentEngine->intersects( &pointXY ) && mZRange.contains( z ) )
```

nyalldawson replied:
(Using prepared geometries for all the containments checks will be MUCH faster)

NEDJIMAbelgacem replied:
Wow that's much faster than my initial implementation especially for circle points. Thanks

NEDJIMAbelgacem committed:
Use geometry engine

NEDJIMAbelgacem committed:
remove render context from the provider code

NEDJIMAbelgacem committed:
remove geometry intersection check from the renderer's traverseTree

nyalldawson commented on src/core/pointcloud/qgspointcloudlayerrenderer.h:
```diff
@@ -75,6 +78,25 @@ class CORE_EXPORT QgsPointCloudLayerRenderer: public QgsMapLayerRenderer
     QgsGeometry mCloudExtent;
 
+    /**
+     * Retrieves the x and y coordinate for the point at index \a i.
```
These two additions are no longer required, right?

nyalldawson commented on src/core/pointcloud/qgspointcloudrenderer.cpp:
```diff
@@ -190,4 +195,71 @@ void QgsPointCloudRenderer::setPointSymbol( PointSymbol symbol )
+      QgsPointXY point1MapCoords = renderContext.mapToPixel().toMapCoordinates( point1.x(), point1.y() );
+      QgsPointXY point2MapCoords = renderContext.mapToPixel().toMapCoordinates( point2.x(), point2.y() );
+      QgsCircle circle = QgsCircle::from2Points( QgsPoint( point1MapCoords ), QgsPoint( point2MapCoords ) );
+      QgsPolygon *polygon = circle.toPolygon( 6 );
+      QgsGeometry circleGeometry( polygon );
```
```suggestion
      std::unique_ptr< QgsPolygon > polygon( circle.toPolygon( 6 ) );
      QgsGeometry circleGeometry( std::move( polygon ) );
```

(avoids the raw pointer, which makes me nervous to look at :smile: )

nyalldawson reviewed (commented):
Brilliant! So close now, just a few minor cleanups left to do. API looks fantastic, thanks.

Can you add some tests for the data provider identify method too? (testqgseptprovider.cpp would be a good place)

NEDJIMAbelgacem committed:
Update src/gui/qgsmaptoolidentify.cpp

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

NEDJIMAbelgacem committed:
Update src/core/pointcloud/qgspointcloudrenderer.h

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

NEDJIMAbelgacem committed:
Update src/core/pointcloud/qgspointcloudrenderer.h

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

NEDJIMAbelgacem committed:
Address reviews

NEDJIMAbelgacem committed:
Added tests

NEDJIMAbelgacem committed:
remove SIP_SKIP and qDebug()

NEDJIMAbelgacem committed:
sipify identify function in dataprovider

NEDJIMAbelgacem commented on src/core/pointcloud/qgspointclouddataprovider.h:
```diff
@@ -73,6 +74,19 @@ class CORE_EXPORT QgsPointCloudDataProvider: public QgsDataProvider
+     * \note this function does not handle elevation properties and you need to
+     * change elevation coordinates yourself after returning from the function
+     */
+    QVector<QMap<QString, QVariant>> identify( float maxErrorInMapCoords, QgsGeometry extentGeometry, const QgsDoubleRange extentZRange, int pointsLimit = 1000 ) SIP_SKIP;
```
I don't like dealing with SIP errors so I just ignored it at first.

NEDJIMAbelgacem commented:
Hi @nyalldawson 
Thanks for the reviews.
I cleaned up according to your suggestions and added tests to the identify function in the dataprovider.
I also removed the SIP_SKIP by adding custom python method code. Does it look good to you?
I tested the python API with this script and it works well:
https://gist.github.com/NEDJIMAbelgacem/cb9fae512532c763b74282d35cedf8db

nyalldawson commented:
Thanks @NEDJIMAbelgacem !
